test/purview_test.tsx

```diff
@@ -1905,6 +1908,72 @@ test("reconnect early", async () => {
   })
 })
 
+test("cleanup happens when 'close' and 'connect' race", async () => {
+  let mountCount = 0
+  let unmountCount = 0
+
+  class Foo extends TestComponent<{}, { text: string }> {
+    state = { text: "hi" }
+
+    constructor(props: {}) {
+      super(props)
+    }
+
+    componentDidMount(): void {
+      mountCount++
+    }
+
+    componentWillUnmount(): void {
+      unmountCount++
+    }
+
+    render(): JSX.Element {
+      return <p>{this.state.text}</p>
+    }
+  }
+
+  await renderAndConnect(<Foo />, async conn => {
+    expect(mountCount).toBe(1)
+    expect(unmountCount).toBe(0)
+    conn.ws.close()
+
+    // Wait for state to be saved and unmount to occur. And wait for the state
+    // to be saved into the db to help simulate the page refresh scenario.
+    await new Promise(resolve => setTimeout(resolve, 25))
+    expect(mountCount).toBe(1)
+    expect(unmountCount).toBe(1)
+
+    const origin = `http://localhost:${conn.port}`
+    const ws = new WebSocket(`ws://localhost:${conn.port}`, { origin })
+    await new Promise(resolve => ws.addEventListener("open", resolve))
+
+    // To avoid infinite recursion, we only mock the first call to
+    // `getStateTree`. To allow jest to cleanup the mock implementation, we use
+    // `spyOn` and call `spy.mockRestore()` once test execution has completed.
+    // This prevent subsequnt tests from using the mock implementation.
+    const spy = jest.spyOn(reloadOptions, "getStateTree")
+    spy.mockImplementationOnce(async rootID => {
+      // Create an artificial delay when deleting state trees so that "connect"
+      // takes a while to let the "close" event execute before "connect"
+      // finishes.
+      await new Promise(resolve => setTimeout(resolve, 200))
+      return await reloadOptions.getStateTree(rootID)
+    })
+    const connect: ClientMessage = {
+      type: "connect",
+      rootIDs: [conn.rootID],
+    }
+    ws.send(JSON.stringify(connect))
+    ws.close()
+
+    // Wait for state to be saved and unmount to occur in "connect."
```

Comment:
unmount to occur in connect is an implementation detail that this test shouldn't really care about - I'd just say wait for state to be saved and unmount to occur.